### PR TITLE
Add ability to define JS modules and Template specific resources

### PR DESCRIPTION
**Pull request by philippjfr**
Closes https://github.com/holoviz/panel/issues/1942
Closes https://github.com/holoviz/panel/issues/1923
Closes https://github.com/holoviz/panel/issues/1124
Closes https://github.com/holoviz/panel/issues/1947 (no longer uses `RenderItem.use_for_title`)

### CODE REVIEW

File: panel/_templates/autoload_panel_js.js [hunk 1]
```diff
@@ -108,10 +108,13 @@ calls it with the rendered model.
       element.onerror = on_error;
       element.async = false;
       element.src = url;
+      if (!url.endsWith('.js')) {
+	element.type = "module"
```
Comment (philippjfr) [inline panel/_templates/autoload_panel_js.js:112]:
```suggestion
      element.type = "module"
```

File: panel/_templates/autoload_panel_js.js [hunk 2]
```diff
@@ -108,10 +108,13 @@ calls it with the rendered model.
       element.onerror = on_error;
       element.async = false;
       element.src = url;
+      if (!url.endsWith('.js')) {
+      element.type = "module"
```
Comment (philippjfr) [inline panel/_templates/autoload_panel_js.js:112]:
```suggestion
        element.type = "module"
```

**Comment by philippjfr** [comment]
@MarcSkovMadsen If you get a chance could you try this PR?

**Comment by philippjfr** [comment]
Okay, JS module handling is now explicit.

File: panel/template/material/material.html [hunk 3]
```diff
@@ -24,6 +23,24 @@
 {% endif %}
 <link href="https://fonts.googleapis.com/css?family=Roboto:300,400,500" rel="stylesheet">
 <link href="https://fonts.googleapis.com/css?family=Material+Icons&display=block" rel="stylesheet">
+{% for src in template_resources['extra_css'] %}
+<link rel="stylesheet"  href="{{ src }}" crossorigin="anonymous"></script>
+
+g{% endfor %}
```
Comment (philippjfr) [inline panel/template/material/material.html:29]:
```suggestion
{% endfor %}
```

**Comment by MarcSkovMadsen** [comment]
> @MarcSkovMadsen If you get a chance could you try this PR? It relies on a maybe faulty assumption that regular JS scripts have `.js` file extensions while modules do not.

Hi @philippjfr . I tried looking at the file changes. But it was not obvious to me what I should test. But for I will take a look at whether there is a global js_modules or template js_modules (or both) I can use.

**Comment by philippjfr** [comment]
>But it was not obvious to me what I should test. But for I will take a look at whether there is a global js_modules or template js_modules (or both) I can use.

Precisely, test the global JS modules (`pn.config.js_modules`) and template specific resources (`template.config.???`).

**Comment by MarcSkovMadsen** [comment]
For the template I 

```bash
pip install bokeh==2.3.0.dev12
panel build panel
```

and `panel serve` this example

```python
import panel as pn

template = pn.template.VanillaTemplate(title="hello")
template.config.js_modules["fast"]="https://unpkg.com/@microsoft/fast-components"

hello = pn.pane.Markdown("""
<fast-design-system-provider use-defaults>
    <fast-button>Hello world</fast-button>
</fast-design-system-provider>
""")
template.main.append("Hello")

template.servable()
```

**I would have expected to see a FastButton. And I see it. Don't know why it's not rounded though.**. **IT WORKS FOR THE TEMPLATE**

image

**Comment by MarcSkovMadsen** [comment]
And for the global configuration it looks like above when I change to this example

```python
import panel as pn
pn.config.js_modules["fast"]="https://unpkg.com/@microsoft/fast-components"

template = pn.template.VanillaTemplate(title="hello")

hello = pn.pane.Markdown("""
<fast-design-system-provider use-defaults>
    <fast-button appearance="accent">Hello world</fast-button>
</fast-design-system-provider>
""")
template.main.append(hello)

template.servable()
```

**Comment by philippjfr** [comment]
Great, not sure why it's not rounded either. Does this cover the features you were hoping for?

**Comment by philippjfr** [comment]
Going to merge. Can follow up with fixes and changes.

**Pull request merged**